File: frontend/src/lib/stores/army.ts
import { writable } from "svelte/store";
import { toast } from "svelte-sonner";
import { api } from "../api/client";
import type { TroopType } from "./troop";

// Mission types matching backend
export type MissionType = 'raid' | 'attack' | 'conquer' | 'support' | 'scout' | 'settle';

export interface CarriedResources {
    wood: number;
    clay: number;
    iron: number;
    crop: number;
}

export type TroopCounts = Partial<Record<TroopType, number>>;

export interface Army {
    id: string;
    player_id: string;
    from_village_id: string;
    to_x: number;
    to_y: number;
    to_village_id: string | null;
    mission: MissionType;
    troops: TroopCounts;
    resources: CarriedResources;
    departed_at: string;
    arrives_at: string;
    returns_at: string | null;
    is_returning: boolean;
    is_stationed: boolean;
}

export interface BattleReport {
    id: string;
    attacker_player_id: string;
    defender_player_id: string | null;
    attacker_village_id: string;
    defender_village_id: string | null;
    mission: MissionType;
    attacker_troops: TroopCounts;
    defender_troops: TroopCounts;
    attacker_losses: TroopCounts;
    defender_losses: TroopCounts;
    resources_stolen: CarriedResources;
    winner: 'attacker' | 'defender' | 'draw';
    occurred_at: string;
    is_read: boolean;
}

export interface ScoutReport {
    id: string;
    attacker_player_id: string;
    defender_player_id: string | null;
    attacker_village_id: string;
    defender_village_id: string | null;
    attacker_scouts: number;
    defender_scouts: number;
    attacker_scouts_lost: number;
    defender_scouts_lost: number;
    success: boolean;
    scouted_resources: CarriedResources | null;
    scouted_troops: TroopCounts | null;
    occurred_at: string;
    is_read: boolean;
}

export interface SendArmyRequest {
    to_x: number;
    to_y: number;
    mission: MissionType;
    troops: TroopCounts;
    resources?: CarriedResources;
}

interface ArmyState {
    outgoingArmies: Army[];
    incomingArmies: Army[];
    stationedTroops: Army[];
    supportSent: Army[];
    reports: BattleReport[];
    scoutReports: ScoutReport[];
    unreadCount: number;
    loading: boolean;
    error: string | null;
}

function createArmyStore() {
    const { subscribe, set, update } = writable<ArmyState>({
        outgoingArmies: [],
        incomingArmies: [],
        stationedTroops: [],
        supportSent: [],
        reports: [],
        scoutReports: [],
        unreadCount: 0,
        loading: false,
        error: null,
    });

    return {
        subscribe,

        // Send army from village
        sendArmy: async (villageId: string, request: SendArmyRequest) => {
            update(state => ({ ...state, loading: true, error: null }));
            try {
                const army = await api.post<Army>(
                    `/api/villages/${villageId}/armies`,
                    request
                );

                update(state => ({
                    ...state,
                    outgoingArmies: [...state.outgoingArmies, army],
                    loading: false,
                }));

                const missionLabel = getMissionLabel(request.mission);
                toast.success('Army Sent', {
                    description: `${missionLabel} to (${request.to_x}, ${request.to_y})`
                });

                return army;
            } catch (error: any) {
                const message = error.message || 'Failed to send army';
                update(state => ({
                    ...state,
                    loading: false,
                    error: message,
                }));
                toast.error('Send Failed', { description: message });
                throw error;
            }
        },

        // Load outgoing armies from village
        loadOutgoing: async (villageId: string) => {
            update(state => ({ ...state, loading: true, error: null }));
            try {
                const armies = await api.get<Army[]>(
                    `/api/villages/${villageId}/armies/outgoing`
                );
                update(state => ({
                    ...state,
                    outgoingArmies: armies,
                    loading: false,
                }));
                return armies;
            } catch (error: any) {
                const message = error.message || 'Failed to load outgoing armies';
                update(state => ({
                    ...state,
                    loading: false,
                    error: message,
                }));
                throw error;
            }
        },

        // Load incoming armies to village
        loadIncoming: async (villageId: string) => {
            update(state => ({ ...state, loading: true, error: null }));
            try {
                const armies = await api.get<Army[]>(
                    `/api/villages/${villageId}/armies/incoming`
                );
                update(state => ({
                    ...state,
                    incomingArmies: armies,
                    loading: false,
                }));
                return armies;
            } catch (error: any) {
                const message = error.message || 'Failed to load incoming armies';
                update(state => ({
                    ...state,
                    loading: false,
                    error: message,
                }));
                throw error;
            }
        },

        // Load battle reports
        loadReports: async () => {
            update(state => ({ ...state, loading: true, error: null }));
            try {
                const reports = await api.get<BattleReport[]>('/api/reports');
                update(state => ({
                    ...state,
                    reports,
                    loading: false,
                }));
                return reports;
            } catch (error: any) {
                const message = error.message || 'Failed to load reports';
                update(state => ({
                    ...state,
                    loading: false,
                    error: message,
                }));
                throw error;
            }
        },

        // Load single report
        loadReport: async (reportId: string) => {
            try {
                return await api.get<BattleReport>(`/api/reports/${reportId}`);
            } catch (error: any) {
                console.error('Failed to load report:', error);
                throw error;
            }
        },

        // Mark report as read
        markRead: async (reportId: string) => {
            try {
                await api.post(`/api/reports/${reportId}/read`, {});
                update(state => ({
                    ...state,
                    reports: state.reports.map(r =>
                        r.id === reportId ? { ...r, is_read: true } : r
                    ),
                    unreadCount: Math.max(0, state.unreadCount - 1),
                }));
            } catch (error: any) {
                console.error('Failed to mark report as read:', error);
            }
        },

        // Load unread count
        loadUnreadCount: async () => {
            try {
                const response = await api.get<{ unread_count: number }>(
                    '/api/reports/unread-count'
                );
                update(state => ({
                    ...state,
                    unreadCount: response.unread_count,
                }));
                return response.unread_count;
            } catch (error: any) {
                console.error('Failed to load unread count:', error);
                return 0;
            }
        },

        // ==================== Scout Reports ====================

        // Load scout reports
        loadScoutReports: async () => {
            update(state => ({ ...state, loading: true, error: null }));
            try {
                const reports = await api.get<ScoutReport[]>('/api/scout-reports');
                update(state => ({
                    ...state,
                    scoutReports: reports,
                    loading: false,
                }));
                return reports;
            } catch (error: any) {
                const message = error.message || 'Failed to load scout reports';
                update(state => ({
                    ...state,
                    loading: false,
                    error: message,
                }));
                throw error;
            }
        },

        // Load single scout report
        loadScoutReport: async (reportId: string) => {
            try {
                return await api.get<ScoutReport>(`/api/scout-reports/${reportId}`);
            } catch (error: any) {
                console.error('Failed to load scout report:', error);
                throw error;
            }
        },

        // Mark scout report as read
        markScoutReportRead: async (reportId: string) => {
            try {
                await api.post(`/api/scout-reports/${reportId}/read`, {});
                update(state => ({
                    ...state,
                    scoutReports: state.scoutReports.map(r =>
                        r.id === reportId ? { ...r, is_read: true } : r
                    ),
                    unreadCount: Math.max(0, state.unreadCount - 1),
                }));
            } catch (error: any) {
                console.error('Failed to mark scout report as read:', error);
            }
        },

        // ==================== Stationed/Support Troops ====================

        // Load troops stationed at a village (support from allies)
        loadStationed: async (villageId: string) => {
            update(state => ({ ...state, loading: true, error: null }));
            try {
                const armies = await api.get<Army[]>(
                    `/api/villages/${villageId}/stationed`
                );
                update(state => ({
                    ...state,
                    stationedTroops: armies,
                    loading: false,
                }));
                return armies;
            } catch (error: any) {
                const message = error.message || 'Failed to load stationed troops';
                update(state => ({
                    ...state,
                    loading: false,
                    error: message,
                }));
                throw error;
            }
        },

        // Load support troops sent to other villages
        loadSupportSent: async () => {
            update(state => ({ ...state, loading: true, error: null }));
            try {
                const armies = await api.get<Army[]>('/api/support-sent');
                update(state => ({
                    ...state,
                    supportSent: armies,
                    loading: false,
                }));
                return armies;
            } catch (error: any) {
                const message = error.message || 'Failed to load support sent';
                update(state => ({
                    ...state,
                    loading: false,
                    error: message,
                }));
                throw error;
            }
        },

        // Recall stationed support troops
        recallSupport: async (armyId: string) => {
            update(state => ({ ...state, loading: true, error: null }));
            try {
                const army = await api.post<Army>(
                    `/api/armies/${armyId}/recall`,
                    {}
                );

                // Remove from supportSent and add to outgoingArmies
                update(state => ({
                    ...state,
                    supportSent: state.supportSent.filter(a => a.id !== armyId),
                    outgoingArmies: [...state.outgoingArmies, army],
                    loading: false,
                }));

                toast.success('Troops Recalled', {
                    description: 'Support troops are returning home'
                });

                return army;
            } catch (error: any) {
                const message = error.message || 'Failed to recall troops';
                update(state => ({
                    ...state,
                    loading: false,
                    error: message,
                }));
                toast.error('Recall Failed', { description: message });
                throw error;
            }
        },

        // Clear error
        clearError: () => {
            update(state => ({ ...state, error: null }));
        },

        // Reset store
        reset: () => {
            set({
                outgoingArmies: [],
                incomingArmies: [],
                stationedTroops: [],
                supportSent: [],
                reports: [],
                scoutReports: [],
                unreadCount: 0,
                loading: false,
                error: null,
            });
        },
    };
}

export const armyStore = createArmyStore();

// Helper functions
export function getMissionLabel(mission: MissionType): string {
    const labels: Record<MissionType, string> = {
        raid: 'Raid',
        attack: 'Attack',
        conquer: 'Conquer',
        support: 'Reinforce',
        scout: 'Scout',
        settle: 'Settle',
    };
    return labels[mission] || mission;
}

export function getMissionIcon(mission: MissionType): string {
    const icons: Record<MissionType, string> = {
        raid: '💰',
        attack: '⚔️',
        conquer: '🏰',
        support: '🛡️',
        scout: '👁️',
        settle: '🏗️',
    };
    return icons[mission] || '📍';
}

export function getMissionColor(mission: MissionType): string {
    const colors: Record<MissionType, string> = {
        raid: 'text-yellow-600',
        attack: 'text-red-600',
        conquer: 'text-purple-600',
        support: 'text-blue-600',
        scout: 'text-gray-600',
        settle: 'text-green-600',
    };
    return colors[mission] || 'text-gray-600';
}

export function getWinnerLabel(winner: string, isAttacker: boolean): string {
    if (winner === 'draw') return 'Draw';
    const won = (winner === 'attacker') === isAttacker;
    return won ? 'Victory' : 'Defeat';
}

export function getWinnerColor(winner: string, isAttacker: boolean): string {
    if (winner === 'draw') return 'text-gray-600';
    const won = (winner === 'attacker') === isAttacker;
    return won ? 'text-green-600' : 'text-red-600';
}

export function formatArrivalTime(arrivedAt: string): string {
    const arrival = new Date(arrivedAt);
    const now = new Date();
    const diff = arrival.getTime() - now.getTime();

    if (diff <= 0) return 'Arrived';

    const hours = Math.floor(diff / (1000 * 60 * 60));
    const minutes = Math.floor((diff % (1000 * 60 * 60)) / (1000 * 60));
    const seconds = Math.floor((diff % (1000 * 60)) / 1000);

    if (hours > 0) {
        return `${hours}h ${minutes}m`;
    }
    if (minutes > 0) {
        return `${minutes}m ${seconds}s`;
    }
    return `${seconds}s`;
}

export function getTotalTroops(troops: TroopCounts): number {
    return Object.values(troops).reduce((sum, count) => sum + (count || 0), 0);
}

export function getTotalResources(resources: CarriedResources): number {
    return resources.wood + resources.clay + resources.iron + resources.crop;
}

File: backend/src/models/army.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use std::collections::HashMap;
use uuid::Uuid;

use super::troop::TroopType;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, sqlx::Type, PartialEq)]
#[sqlx(type_name = "mission_type", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum MissionType {
    Raid,     // Quick attack, steal resources
    Attack,   // Full attack, kill troops
    Conquer,  // Attack to take over village
    Support,  // Send troops to defend
    Scout,    // Reconnaissance mission
    Settle,   // Found new village with settlers
}

impl MissionType {
    pub fn is_hostile(&self) -> bool {
        matches!(self, MissionType::Raid | MissionType::Attack | MissionType::Conquer | MissionType::Scout)
    }

    pub fn is_support(&self) -> bool {
        matches!(self, MissionType::Support)
    }

    pub fn returns(&self) -> bool {
        // Settle missions don't return
        !matches!(self, MissionType::Settle)
    }
}

/// Troops in an army (serialized as JSON in database)
pub type ArmyTroops = HashMap<TroopType, i32>;

/// Resources carried by army (serialized as JSON in database)
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct CarriedResources {
    #[serde(default)]
    pub wood: i32,
    #[serde(default)]
    pub clay: i32,
    #[serde(default)]
    pub iron: i32,
    #[serde(default)]
    pub crop: i32,
}

impl CarriedResources {
    pub fn total(&self) -> i32 {
        self.wood + self.clay + self.iron + self.crop
    }
}

/// Army movement record
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Army {
    pub id: Uuid,
    pub player_id: Uuid,
    pub from_village_id: Uuid,
    pub to_x: i32,
    pub to_y: i32,
    pub to_village_id: Option<Uuid>,
    pub mission: MissionType,
    pub troops: sqlx::types::Json<ArmyTroops>,
    pub resources: sqlx::types::Json<CarriedResources>,
    pub departed_at: DateTime<Utc>,
    pub arrives_at: DateTime<Utc>,
    pub returns_at: Option<DateTime<Utc>>,
    pub is_returning: bool,
    pub battle_report_id: Option<Uuid>,
    pub created_at: DateTime<Utc>,
}

/// Battle report record
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct BattleReport {
    pub id: Uuid,
    pub attacker_player_id: Uuid,
    pub defender_player_id: Option<Uuid>,
    pub attacker_village_id: Uuid,
    pub defender_village_id: Option<Uuid>,
    pub mission: MissionType,
    pub attacker_troops: sqlx::types::Json<ArmyTroops>,
    pub defender_troops: sqlx::types::Json<ArmyTroops>,
    pub attacker_losses: sqlx::types::Json<ArmyTroops>,
    pub defender_losses: sqlx::types::Json<ArmyTroops>,
    pub resources_stolen: sqlx::types::Json<CarriedResources>,
    pub winner: String, // "attacker", "defender", "draw"
    pub occurred_at: DateTime<Utc>,
    pub read_by_attacker: bool,
    pub read_by_defender: bool,
    pub created_at: DateTime<Utc>,
}

// Request/Response DTOs

#[derive(Debug, Clone, Deserialize)]
pub struct SendArmyRequest {
    pub to_x: i32,
    pub to_y: i32,
    pub mission: MissionType,
    pub troops: HashMap<TroopType, i32>,
    #[serde(default)]
    pub resources: CarriedResources,
}

#[derive(Debug, Clone, Serialize)]
pub struct ArmyResponse {
    pub id: Uuid,
    pub from_village_id: Uuid,
    pub to_x: i32,
    pub to_y: i32,
    pub to_village_id: Option<Uuid>,
    pub mission: MissionType,
    pub troops: ArmyTroops,
    pub resources: CarriedResources,
    pub departed_at: DateTime<Utc>,
    pub arrives_at: DateTime<Utc>,
    pub returns_at: Option<DateTime<Utc>>,
    pub is_returning: bool,
}

impl From<Army> for ArmyResponse {
    fn from(a: Army) -> Self {
        Self {
            id: a.id,
            from_village_id: a.from_village_id,
            to_x: a.to_x,
            to_y: a.to_y,
            to_village_id: a.to_village_id,
            mission: a.mission,
            troops: a.troops.0,
            resources: a.resources.0,
            departed_at: a.departed_at,
            arrives_at: a.arrives_at,
            returns_at: a.returns_at,
            is_returning: a.is_returning,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct BattleReportResponse {
    pub id: Uuid,
    pub attacker_player_id: Uuid,
    pub defender_player_id: Option<Uuid>,
    pub attacker_village_id: Uuid,
    pub defender_village_id: Option<Uuid>,
    pub mission: MissionType,
    pub attacker_troops: ArmyTroops,
    pub defender_troops: ArmyTroops,
    pub attacker_losses: ArmyTroops,
    pub defender_losses: ArmyTroops,
    pub resources_stolen: CarriedResources,
    pub winner: String,
    pub occurred_at: DateTime<Utc>,
    pub is_read: bool,
}

impl BattleReport {
    pub fn to_response(&self, is_attacker: bool) -> BattleReportResponse {
        BattleReportResponse {
            id: self.id,
            attacker_player_id: self.attacker_player_id,
            defender_player_id: self.defender_player_id,
            attacker_village_id: self.attacker_village_id,
            defender_village_id: self.defender_village_id,
            mission: self.mission,
            attacker_troops: self.attacker_troops.0.clone(),
            defender_troops: self.defender_troops.0.clone(),
            attacker_losses: self.attacker_losses.0.clone(),
            defender_losses: self.defender_losses.0.clone(),
            resources_stolen: self.resources_stolen.0.clone(),
            winner: self.winner.clone(),
            occurred_at: self.occurred_at,
            is_read: if is_attacker { self.read_by_attacker } else { self.read_by_defender },
        }
    }
}
